File: Online-judge/leetcode/linked-list-random-node.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/linked-list-random-node/description/
 */

/**
 * @brief 
 * Time complexity : O(n)
 * Space complexity: O(1)
 */
 


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
private:
    ListNode* mHead;
public:
    Solution(ListNode* head) {
        mHead = head;
    }
    
    int getRandom() {
        ListNode* curr = mHead;
        int res = 0, i = 1;
        while (curr) {
            if (rand() % i == 0) {
                res = curr->val;
            }
            i++;
            curr = curr->next;
        }
        return res;
    }
};

File: Online-judge/leetcode/linked-list-random-node.py

import random

'''
Time complexity: O(n)
Space Complexity: O(n)
'''

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.values = []
        while head:
            self.values.append(head.val)
            head = head.next

    def getRandom(self) -> int:
        index = int(random.random() * len(self.values))
        return self.values[index]


# Your Solution object will be instantiated and called as such:
# obj = Solution(head)
# param_1 = obj.getRandom()